internal/app/networkd/pkg/nic/nic.go

```diff
@@ -2,75 +2,323 @@
+			return err
+		}
+		// Surely there's a better way to do a *uint32
+		var bondIndex *uint32
+		bondIndex = new(uint32)
+		*bondIndex = uint32(n.Link.Index)
+
+		if err = n.enslaveLink(bondIndex, n.SubInterfaces...); err != nil {
+			return err
+		}
+	}
+
+	log.Println("bringing up", n.Link.Name)
+	if err = n.rtnlConn.LinkUp(n.Link); err != nil {
+		log.Printf("failed to bring up %s: %v", n.Name, err)
+		return err
+	}
+
+	// Wait for link to report up
+	err = retry.Exponential(30*time.Second, retry.WithUnits(250*time.Millisecond), retry.WithJitter(50*time.Millisecond)).Retry(func() error {
+		link, err := n.rtnlConn.LinkByIndex(n.Link.Index)
+		if err != nil {
+			retry.UnexpectedError(err)
+		}
+		if link.Flags&net.FlagUp != net.FlagUp {
+			return retry.ExpectedError(fmt.Errorf("link is not up %s", n.Link.Name))
+		}
+
+		log.Println("link is up", link.Flags.String())
+		return nil
+	})
+
+	if err != nil {
+		return fmt.Errorf("failed to bring up interface%s: %w", n.Link.Name, err)
+	}
+
+	return err
+}
+
+func (n *NetworkInterface) Addressing() error {
+	for _, method := range n.AddressMethod {
+		log.Printf("configuring %s addressing for %s\n", method.Name(), n.Name)
+
+		if err := n.configureInterface(method); err != nil {
+			// Treat as non fatal error when failing to configure an interface
+			log.Println(err)
+			return nil
+		}
+
+		if !method.Valid() {
+			return nil
+		}
+	}
+
+	return nil
+}
+
+func (n *NetworkInterface) Renew() {
+	for _, method := range n.AddressMethod {
+		if method.TTL() == 0 {
+			continue
+		}
+
+		go n.renew(method)
+	}
+}
+
+// renew sets up the looping to ensure we keep the addressing information
+// up to date. We attempt to do our first reconfiguration halfway through
+// address TTL. If that fails, we'll continue to attempt to retry every
+// halflife.
+func (n *NetworkInterface) renew(method address.Addressing) {
+	renewDuration := method.TTL() / 2
+
+	var err error
+
+	for {
+		<-time.After(renewDuration)
+
+		if err = n.configureInterface(method); err != nil {
+			log.Printf("failed to renew interface address for %s: %v\n", method.Link().Name, err)
+
+			renewDuration = (renewDuration / 2)
+		} else {
+			renewDuration = method.TTL() / 2
+		}
+	}
+}
+
+// address handles the actual address discovery mechanism and
+// netlink interaction to configure the interface
+// nolint: gocyclo
+func (n *NetworkInterface) configureInterface(method address.Addressing) error {
+	// TODO s/Discover/Something else/
+	// TODO make context more relevant
+	var err error
+	if err = method.Discover(context.Background(), n.Link); err != nil {
+		// Right now this would only happen during dhcp discovery failure
+		log.Printf("failed to prep %s: %v", method.Link().Name, err)
+		return err
+	}
+
+	// Set link MTU if we got a response
+	if err = n.setMTU(method.Link().Index, method.MTU()); err != nil {
+		log.Printf("failed to set mtu %d for %s: %v", method.MTU(), method.Link().Name, err)
```

Comment:
Remove the logging. Should be up to consumer to log. We can `return fmt.Errorf` this line instead.